frontend/src/utils/date.ts

```diff
@@ -62,3 +71,46 @@ export const isDateOutOfRange = (
 
   return false
 }
+
+export const ISO_DATE_FORMAT_REGEX =
+  /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-](\d+(:\d+)?))?$/
+
+export const isIsoDateString = (value: unknown): value is JsonDate => {
+  return (
+    typeof value === 'string' &&
+    ISO_DATE_FORMAT_REGEX.test(value) &&
+    !isNaN(new Date(value).getTime())
+  )
+}
+
+/**
+ * This function mutates given @param body, and transforms all ISO date strings
+ * in the body object to Date objects.
+ * @param body object to transform
+ */
+export const mutableTransformAllIsoStringsToDate = (body: unknown) => {
+  if (body === null || body === undefined || typeof body !== 'object') {
+    return
+  }
+
+  for (const key of Object.keys(body)) {
+    const value = (body as Record<string, unknown>)[key]
+    if (isIsoDateString(value)) {
+      // eslint-disable-next-line @typescript-eslint/no-extra-semi
+      ;(body as Record<string, unknown>)[key] = parseISO(value)
+    } else if (typeof value === 'object') {
+      mutableTransformAllIsoStringsToDate(value)
+    }
+  }
+}
+
+/**
+ * Non-mutable version of `mutableTransformAllIsoStringsToDate`
+ * @param body object to transform
+ * @returns transformed object with all ISO date strings transformed to Date objects.
+ */
+export const transformAllIsoStringsToDate = <T>(body: T): T => {
+  const result = { ...body }
+  mutableTransformAllIsoStringsToDate(result)
+  return result
```

Comment:
Warning on this: the shallow copy makes it look like there's no mutation, but `mutableTransformAllIsoStringsToDate()` is a recursive function, so when it descends into nested objects, it *does* mutate the input.

In

Reply:
I'm thinking that given the use cases, it's OK to let `transformAllIsoStringsToDate()` be a function that mutates btw 🤔 , so perhaps we can just change this to be:
```suggestion
  mutableTransformAllIsoStringsToDate(body)
  return body
```

Reply:
Done in [309af6d](https://github.com/opengovsg/FormSG/pull/3571/commits/309af6dbe40e109e13b0de11ef851e5916b00eee)